Make ReadOperation and WriteOperation thread-safe

szewai opened the pull request:
#### 6aa2a15a3955ac4f4107074b06b661493ff0aced
<pre>
Make ReadOperation and WriteOperation thread-safe
<a href="https://bugs.webkit.org/show_bug.cgi?id=282519">https://bugs.webkit.org/show_bug.cgi?id=282519</a>
<a href="https://rdar.apple.com/139207292">rdar://139207292</a>

Reviewed by NOBODY (OOPS!).

In NetworkCache::Storage, ReadOperation and WriteOperation can be accessed from different threads, and some of their
data members are read and written on different threads, which could lead to data race. To fix this, this patch makes a
few changes:
1. Add threading annotation to data members of ReadOperation and WriteOperation.
2. Make ReadOperation and WriteOperation class instead of struct, and their data members can only be accessed via public
functions which acquires lock or performs threading check as needed.
3. For data members that could contain AtomString, like NetworkCache::Key and NetworkCache::Storage::Record, make
ReadOperation and WriteOperation either store isolated copy of them, or ensure they are only accessed from main thread
and make isolated copy of them when passing them across thread.
4. Make ReadOperation directly stores Storage::Record like WriteOperation does, instead of storing a pointer to it.

This patch also makes some improvments, including removing unnecessary data members, on NetworkCache::IOChannel to make
it more safe, as it is also designed to be accessed from multiple threads (ConcurrentWorkQueue).

* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::Cache::retrieve):
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannel.h:
(WebKit::NetworkCache::IOChannel::open):
(WebKit::NetworkCache::IOChannel::WTF_GUARDED_BY_LOCK):
(WebKit::NetworkCache::IOChannel::path const): Deleted.
(WebKit::NetworkCache::IOChannel::type const): Deleted.
(WebKit::NetworkCache::IOChannel::isOpened const): Deleted.
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelCocoa.mm:
(WebKit::NetworkCache::IOChannel::IOChannel):
(WebKit::NetworkCache::IOChannel::read):
(WebKit::NetworkCache::IOChannel::write):
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelCurl.cpp:
(WebKit::NetworkCache::IOChannel::IOChannel):
(WebKit::NetworkCache::IOChannel::read):
(WebKit::NetworkCache::IOChannel::write):
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelGLib.cpp:
(WebKit::NetworkCache::IOChannel::IOChannel):
(WebKit::NetworkCache::IOChannel::read):
(WebKit::NetworkCache::IOChannel::write):
* Source/WebKit/NetworkProcess/cache/NetworkCacheKey.h:
(WebKit::NetworkCache::Key::isolatedCopy const):
* Source/WebKit/NetworkProcess/cache/NetworkCacheSpeculativeLoadManager.cpp:
(WebKit::NetworkCache::SpeculativeLoadManager::retrieveEntryFromStorage):
(WebKit::NetworkCache::SpeculativeLoadManager::retrieveSubresourcesEntry):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
(WebKit::NetworkCache::Storage::ReadOperation::ReadOperation):
(WebKit::NetworkCache::Storage::ReadOperation::WTF_GUARDED_BY_CAPABILITY):
(WebKit::NetworkCache::Storage::isHigherPriority):
(WebKit::NetworkCache::Storage::ReadOperation::cancel):
(WebKit::NetworkCache::Storage::ReadOperation::finish):
(WebKit::NetworkCache::Storage::ReadOperation::key const):
(WebKit::NetworkCache::Storage::ReadOperation::ordinal const):
(WebKit::NetworkCache::Storage::ReadOperation::priority const):
(WebKit::NetworkCache::Storage::ReadOperation::isCanceled const):
(WebKit::NetworkCache::Storage::ReadOperation::timings):
(WebKit::NetworkCache::Storage::ReadOperation::setTimings):
(WebKit::NetworkCache::Storage::ReadOperation::setBlob):
(WebKit::NetworkCache::Storage::ReadOperation::setResultRecord):
(WebKit::NetworkCache::Storage::WriteOperation::WriteOperation):
(WebKit::NetworkCache::Storage::WriteOperation::callMappedBodyHandler):
(WebKit::NetworkCache::Storage::WriteOperation::callCompletionHandler):
(WebKit::NetworkCache::Storage::WriteOperation::record const):
(WebKit::NetworkCache::Storage::readRecord):
(WebKit::NetworkCache::Storage::storeBodyAsBlob):
(WebKit::NetworkCache::Storage::removeFromPendingWriteOperations):
(WebKit::NetworkCache::Storage::dispatchReadOperation):
(WebKit::NetworkCache::Storage::finishReadOperation):
(WebKit::NetworkCache::retrieveFromMemory):
(WebKit::NetworkCache::Storage::dispatchWriteOperation):
(WebKit::NetworkCache::Storage::finishWriteOperation):
(WebKit::NetworkCache::Storage::retrieve):
(WebKit::NetworkCache::Storage::traverseWithinRootPath):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:
(WebKit::NetworkCache::Storage::Record::Record):
(WebKit::NetworkCache::Storage::Record::isolatedCopy const):
(WebKit::NetworkCache::Storage::Record::isNull const):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6aa2a15a3955ac4f4107074b06b661493ff0aced

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74826 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54256 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27643 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79250 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26064 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63392 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2041 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58776 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17058 "21 failures") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77893 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48929 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64304 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39168 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46207 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24397 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67375 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22146 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80740 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2144 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1293 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67032 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2292 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64322 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66325 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10272 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8423 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2109 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4896 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2137 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3058 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2144 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelCocoa.mm:
```diff
@@ -96,29 +97,34 @@ static long dispatchQueueIdentifier(WorkQueue::QOS qos)
+
     bool didCallCompletionHandler = false;
-    dispatch_io_read(m_dispatchIO.get(), offset, size, queue.dispatchQueue(), makeBlockPtr([channel, completionHandler = WTFMove(completionHandler), didCallCompletionHandler](bool done, dispatch_data_t fileData, int error) mutable {
+    dispatch_io_read(m_dispatchIO.get(), offset, size, queue->dispatchQueue(), makeBlockPtr([protectedThis = Ref { *this }, queue = WTFMove(queue), completionHandler = WTFMove(completionHandler), didCallCompletionHandler](bool done, dispatch_data_t fileData, int error) mutable {
```
in C++, the order of evaluation of function parameters is unspecified. This means that the compiler is free to evaluate the arguments in any order it chooses. Therefore, I think the new code is likely unsafe with regards to `queue` which is used and moved here.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelCocoa.mm:
```diff
@@ -96,29 +97,34 @@ static long dispatchQueueIdentifier(WorkQueue::QOS qos)
-        ASSERT_UNUSED(done, done);
-        auto callback = WTFMove(completionHandler);
-        callback(error);
+    dispatch_io_write(m_dispatchIO.get(), offset, dispatchData, queue->dispatchQueue(), makeBlockPtr([protectedThis = Ref { *this }, queue = WTFMove(queue), completionHandler = WTFMove(completionHandler)](bool done, dispatch_data_t, int error) mutable {
```
ditto.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
```diff
@@ -69,91 +69,198 @@ static uint64_t nextReadOperationOrdinal()
-    SHA1::Digest expectedBodyHash;
-    BlobStorage::Blob resultBodyBlob;
+    bool isCanceled() const;
+    const Key key() const;
```
seems like this should return a `const Key&` or a `Key`? But likely not a `const Key`

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:
```diff
@@ -53,13 +53,23 @@ class Storage : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Storage,
     static RefPtr<Storage> open(const String& cachePath, Mode, size_t capacity);
 
     struct Record {
+        Record isolatedCopy() const & { return { crossThreadCopy(key), timeStamp, header, body, bodyHash }; }
```
Should probably come *after* the constructors.

cdumez reviewed (commented):
I am curious to see the TSAN output that caused you to write such a massive change and whether or not this could be solved with a more targeted fix.

szewai committed:
Make ReadOperation and WriteOperation thread-safe
https://bugs.webkit.org/show_bug.cgi?id=282519
rdar://139207292

Reviewed by NOBODY (OOPS!).

In NetworkCache::Storage, ReadOperation and WriteOperation can be accessed from different threads, and some of their
data members are read and written on different threads, which could lead to data race. To fix this, this patch makes a
few changes:
1. Add threading annotation to data members of ReadOperation and WriteOperation.
2. Make ReadOperation and WriteOperation class instead of struct, and their data members can only be accessed via public
functions which acquires lock or performs threading check as needed.
3. For data members that could contain AtomString, like NetworkCache::Key and NetworkCache::Storage::Record, make
ReadOperation and WriteOperation either store isolated copy of them, or ensure they are only accessed from main thread
and make isolated copy of them when passing them across thread.
4. Make ReadOperation directly stores Storage::Record like WriteOperation does, instead of storing a pointer to it.

This patch also makes some improvments, including removing unnecessary data members, on NetworkCache::IOChannel to make
it more safe, as it is also designed to be accessed from multiple threads (ConcurrentWorkQueue).

* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::Cache::retrieve):
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannel.h:
(WebKit::NetworkCache::IOChannel::open):
(WebKit::NetworkCache::IOChannel::WTF_GUARDED_BY_LOCK):
(WebKit::NetworkCache::IOChannel::path const): Deleted.
(WebKit::NetworkCache::IOChannel::type const): Deleted.
(WebKit::NetworkCache::IOChannel::isOpened const): Deleted.
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelCocoa.mm:
(WebKit::NetworkCache::IOChannel::IOChannel):
(WebKit::NetworkCache::IOChannel::read):
(WebKit::NetworkCache::IOChannel::write):
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelCurl.cpp:
(WebKit::NetworkCache::IOChannel::IOChannel):
(WebKit::NetworkCache::IOChannel::read):
(WebKit::NetworkCache::IOChannel::write):
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelGLib.cpp:
(WebKit::NetworkCache::IOChannel::IOChannel):
(WebKit::NetworkCache::IOChannel::read):
(WebKit::NetworkCache::IOChannel::write):
* Source/WebKit/NetworkProcess/cache/NetworkCacheKey.h:
(WebKit::NetworkCache::Key::isolatedCopy const):
* Source/WebKit/NetworkProcess/cache/NetworkCacheSpeculativeLoadManager.cpp:
(WebKit::NetworkCache::SpeculativeLoadManager::retrieveEntryFromStorage):
(WebKit::NetworkCache::SpeculativeLoadManager::retrieveSubresourcesEntry):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.cpp:
(WebKit::NetworkCache::Storage::ReadOperation::ReadOperation):
(WebKit::NetworkCache::Storage::ReadOperation::WTF_GUARDED_BY_CAPABILITY):
(WebKit::NetworkCache::Storage::isHigherPriority):
(WebKit::NetworkCache::Storage::ReadOperation::cancel):
(WebKit::NetworkCache::Storage::ReadOperation::finish):
(WebKit::NetworkCache::Storage::ReadOperation::key const):
(WebKit::NetworkCache::Storage::ReadOperation::ordinal const):
(WebKit::NetworkCache::Storage::ReadOperation::priority const):
(WebKit::NetworkCache::Storage::ReadOperation::isCanceled const):
(WebKit::NetworkCache::Storage::ReadOperation::timings):
(WebKit::NetworkCache::Storage::ReadOperation::setTimings):
(WebKit::NetworkCache::Storage::ReadOperation::setBlob):
(WebKit::NetworkCache::Storage::ReadOperation::setResultRecord):
(WebKit::NetworkCache::Storage::WriteOperation::WriteOperation):
(WebKit::NetworkCache::Storage::WriteOperation::callMappedBodyHandler):
(WebKit::NetworkCache::Storage::WriteOperation::callCompletionHandler):
(WebKit::NetworkCache::Storage::WriteOperation::record const):
(WebKit::NetworkCache::Storage::readRecord):
(WebKit::NetworkCache::Storage::storeBodyAsBlob):
(WebKit::NetworkCache::Storage::removeFromPendingWriteOperations):
(WebKit::NetworkCache::Storage::dispatchReadOperation):
(WebKit::NetworkCache::Storage::finishReadOperation):
(WebKit::NetworkCache::retrieveFromMemory):
(WebKit::NetworkCache::Storage::dispatchWriteOperation):
(WebKit::NetworkCache::Storage::finishWriteOperation):
(WebKit::NetworkCache::Storage::retrieve):
(WebKit::NetworkCache::Storage::traverseWithinRootPath):
* Source/WebKit/NetworkProcess/cache/NetworkCacheStorage.h:
(WebKit::NetworkCache::Storage::Record::Record):
(WebKit::NetworkCache::Storage::Record::isolatedCopy const):
(WebKit::NetworkCache::Storage::Record::isNull const):
